### PR TITLE
Replace DefineOpaqueTypes::No => DefineOpaqueTypes::Yes

### LINKED ISSUE #116652: go through uses of `DefineOpaqueTypes::No` and either document or change them
This issue can be resolved in individual steps (each application of all steps should be their own commit!):

1. flip a site to `Yes`
2. see what tests fail, document the site by mentioning the test in a comment.
3. If no tests fail, open a PR and assign to @oli-obk or @compiler-errors and leave a comment that there's no test exercising it. We'll try to come up with a test and/or run crater to find examples.

### PR DESCRIPTION
**Pull request by chansuke**
This PR is part of #116652

### CODE REVIEW

**Comment by compiler-errors** [comment]
Please follow the instructions in #116652. Specifically, copied from that issue:

> (each application of all steps should be their own commit!):
> 1. flip a single usage of `DefineOpauqeTypes::No` to `Yes`
> 2. see what tests fail, document the site by mentioning the test in a comment, revert to `::No`, and open a PR.
> 3. If no tests fail, still open a PR and assign to @oli-obk or @compiler-errors and leave a comment that there's no test exercising it. We'll try to come up with a test and/or run crater to find examples.

@rustbot author

File: compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical.rs [hunk 1]
```diff
@@ -323,7 +323,7 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
             nested_goals.extend(
                 infcx
                     .at(&ObligationCause::dummy(), param_env)
-                    .eq(DefineOpaqueTypes::No, orig, response)
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/solve/eval_ctxt/canonical.rs:326]: Ignore all of the call sites in `rustc_trait_selection/src/solve`. Those can be commented saying that the new trait solver ignores `DefineOpaqueTypes`.

File: compiler/rustc_hir_analysis/src/coherence/builtin.rs [hunk 2]
```diff
@@ -165,7 +165,7 @@ fn visit_implementation_of_dispatch_from_dyn(tcx: TyCtxt<'_>, impl_did: LocalDef
     use rustc_type_ir::TyKind::*;
     match (source.kind(), target.kind()) {
         (&Ref(r_a, _, mutbl_a), Ref(r_b, _, mutbl_b))
-            if infcx.at(&cause, param_env).eq(DefineOpaqueTypes::No, r_a, *r_b).is_ok()
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/coherence/builtin.rs:168]: This needs a FIXME. It should be split out into a separate commit and we need to think about how to create a test for this.

File: compiler/rustc_hir_analysis/src/coherence/builtin.rs [hunk 3]
```diff
@@ -204,7 +204,7 @@ fn visit_implementation_of_dispatch_from_dyn(tcx: TyCtxt<'_>, impl_did: LocalDef
                     }
 
                     if let Ok(ok) =
-                        infcx.at(&cause, param_env).eq(DefineOpaqueTypes::No, ty_a, ty_b)
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/coherence/builtin.rs:207]: Same here and below.

File: compiler/rustc_hir_typeck/src/callee.rs [hunk 4]
```diff
@@ -799,7 +799,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
         let param = callee_args.const_at(host_effect_index);
         let cause = self.misc(span);
-        match self.at(&cause, self.param_env).eq(infer::DefineOpaqueTypes::No, effect, param) {
+        match self.at(&cause, self.param_env).eq(infer::DefineOpaqueTypes::Yes, effect, param) {
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/callee.rs:802]: This needs a FIXME for a test. I would prefer it gets separated into another PR so we can work on creating a test.

**Comment by compiler-errors** [comment]
Specifically, this should be broken up into several PRs:

1. All of the comments that specify which tests fail when changed to `Yes` can be split into a separate PR
2. For all the call sites that don't have tests, they should be split into separate PRs so we can work on creating a test. Or ideally, we'd just open a new issue calling for someone creating a test.

File: compiler/rustc_hir_typeck/src/coercion.rs [hunk 5]
```diff
@@ -1140,7 +1140,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         // are the same function and their parameters have a LUB.
                         match self.commit_if_ok(|_| {
                             self.at(cause, self.param_env).lub(
-                                DefineOpaqueTypes::No,
+                                DefineOpaqueTypes::Yes,
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/coercion.rs:1143]: This needs a test for it. I think we could create a test exercising this by coercing closures to fn ptrs when they have no captures.

File: compiler/rustc_hir_typeck/src/coercion.rs [hunk 6]
```diff
@@ -1194,7 +1194,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             let sig = self
                 .at(cause, self.param_env)
                 .trace(prev_ty, new_ty)
-                .lub(DefineOpaqueTypes::No, a_sig, b_sig)
+                .lub(DefineOpaqueTypes::Yes, a_sig, b_sig)
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/coercion.rs:1197]: Same here. We need to find out what coercions to trigger this.

File: compiler/rustc_hir_typeck/src/coercion.rs [hunk 7]
```diff
@@ -1283,7 +1283,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
                 return self
                     .commit_if_ok(|_| {
-                        self.at(cause, self.param_env).lub(DefineOpaqueTypes::No, prev_ty, new_ty)
+                        self.at(cause, self.param_env).lub(DefineOpaqueTypes::Yes, prev_ty, new_ty)
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/coercion.rs:1286]: Same here and below.

File: compiler/rustc_hir_typeck/src/demand.rs [hunk 8]
```diff
@@ -400,7 +400,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     // what our ideal rcvr ty would look like.
                     let _ = self
                         .at(&ObligationCause::dummy(), self.param_env)
-                        .eq(DefineOpaqueTypes::No, method.sig.inputs()[idx + 1], arg_ty)
+                        .eq(DefineOpaqueTypes::Yes, method.sig.inputs()[idx + 1], arg_ty)
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/demand.rs:403]: This is diagnostics code. This can be set to yes and commented that it's diagnostics code...

File: compiler/rustc_hir_typeck/src/demand.rs [hunk 9]
```diff
@@ -479,7 +479,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     .and_then(|method| {
                         let _ = self
                             .at(&ObligationCause::dummy(), self.param_env)
-                            .eq(DefineOpaqueTypes::No, ideal_rcvr_ty, expected_ty)
+                            .eq(DefineOpaqueTypes::Yes, ideal_rcvr_ty, expected_ty)
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/demand.rs:482]: Same here, this is diagnostics code.

File: compiler/rustc_hir_typeck/src/expr.rs [hunk 10]
```diff
@@ -1746,7 +1746,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                 let target_ty = self.field_ty(base_expr.span, f, args);
                                 let cause = self.misc(base_expr.span);
                                 match self.at(&cause, self.param_env).sup(
-                                    DefineOpaqueTypes::No,
+                                    DefineOpaqueTypes::Yes,
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/expr.rs:1749]: This needs `#![feature(type_changing_struct_update)]`. I think a test could be crafted using that.

File: compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs [hunk 11]
```diff
@@ -542,7 +542,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             let span = self.tcx.hir().body(body_id).value.span;
             let ok = self
                 .at(&self.misc(span), self.param_env)
-                .eq(DefineOpaqueTypes::No, interior, witness)
+                .eq(DefineOpaqueTypes::Yes, interior, witness)
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs:545]: We can keep this yes, since we never expect this to be anything other than equating a ty variable to a concrete type.

File: compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs [hunk 12]
```diff
@@ -1423,7 +1423,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             let impl_ty = self.normalize(span, tcx.type_of(impl_def_id).instantiate(tcx, args));
             let self_ty = self.normalize(span, self_ty);
             match self.at(&self.misc(span), self.param_env).eq(
-                DefineOpaqueTypes::No,
+                DefineOpaqueTypes::Yes,
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs:1426]: I don't know how we can create a test for this... This needs to be a FIXME and we need to find out how to make a test.

File: src/librustdoc/clean/blanket_impl.rs [hunk 13]
```diff
@@ -48,7 +48,7 @@ impl<'a, 'tcx> BlanketImplFinder<'a, 'tcx> {
                 // Require the type the impl is implemented on to match
                 // our type, and ignore the impl if there was a mismatch.
                 let Ok(eq_result) = infcx.at(&traits::ObligationCause::dummy(), param_env).eq(
-                    DefineOpaqueTypes::No,
+                    DefineOpaqueTypes::Yes,
```
Comment (compiler-errors) [inline src/librustdoc/clean/blanket_impl.rs:51]: This is using a `DefiningAnchor::Error`, so this has no effect. leave a comment saying this.

File: compiler/rustc_trait_selection/src/traits/specialize/mod.rs [hunk 14]
```diff
@@ -223,7 +223,7 @@ fn fulfill_implication<'tcx>(
     // do the impls unify? If not, no specialization.
     let Ok(InferOk { obligations: more_obligations, .. }) = infcx
         .at(&ObligationCause::dummy(), param_env)
-        .eq(DefineOpaqueTypes::No, source_trait, target_trait)
+        .eq(DefineOpaqueTypes::Yes, source_trait, target_trait)
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/specialize/mod.rs:226]: This makes me pretty nervous. We need a specialization-based test here.

File: compiler/rustc_trait_selection/src/traits/select/mod.rs [hunk 15]
```diff
@@ -2617,6 +2623,7 @@ impl<'tcx> SelectionContext<'_, 'tcx> {
     ) -> Result<Vec<PredicateObligation<'tcx>>, ()> {
         self.infcx
             .at(&obligation.cause, obligation.param_env)
+            // ICEs if we use `DefineOpaqueTypes::Yes` here, see #116652.
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:2626]: What *kind* of ICE? Where? There's no need to link #116652, but there does need to be more information here.

File: compiler/rustc_trait_selection/src/traits/select/mod.rs [hunk 16]
```diff
@@ -2488,6 +2489,11 @@ impl<'tcx> SelectionContext<'_, 'tcx> {
         let InferOk { obligations, .. } = self
             .infcx
             .at(&cause, obligation.param_env)
+            // #116652
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:2492]: Instead of linking all of these issues with `#116652`, can you just put `FIXME(DefineOpaqueTypes):`? That may be easier.

File: compiler/rustc_trait_selection/src/traits/select/mod.rs [hunk 17]
```diff
@@ -1772,6 +1772,7 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         let is_match = self
             .infcx
             .at(&obligation.cause, obligation.param_env)
+            // ICEs if we use `DefineOpaqueTypes::Yes` here, see #116652.
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:1775]: What kind of ICE?

File: compiler/rustc_trait_selection/src/traits/select/mod.rs [hunk 18]
```diff
@@ -1710,7 +1710,7 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
         });
         self.infcx
             .at(&obligation.cause, obligation.param_env)
-            .sup(DefineOpaqueTypes::No, ty::Binder::dummy(placeholder_trait_ref), trait_bound)
+            .sup(DefineOpaqueTypes::Yes, ty::Binder::dummy(placeholder_trait_ref), trait_bound)
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:1713]: This needs a FIXME that a test should be constructed

File: compiler/rustc_trait_selection/src/traits/select/mod.rs [hunk 19]
```diff
@@ -937,7 +937,7 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                                     .infcx
                                     .at(&obligation.cause, obligation.param_env)
                                     .trace(c1, c2)
-                                    .eq(DefineOpaqueTypes::No, a.args, b.args)
+                                    .eq(DefineOpaqueTypes::Yes, a.args, b.args)
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/select/mod.rs:940]: All of the `::Yes` in this file need to have a FIXME, probably opened with a new PR...

File: compiler/rustc_trait_selection/src/traits/fulfill.rs [hunk 20]
```diff
@@ -631,7 +631,7 @@ impl<'a, 'tcx> ObligationProcessor for FulfillProcessor<'a, 'tcx> {
                 }
                 ty::PredicateKind::Clause(ty::ClauseKind::ConstArgHasType(ct, ty)) => {
                     match self.selcx.infcx.at(&obligation.cause, obligation.param_env).eq(
-                        DefineOpaqueTypes::No,
+                        DefineOpaqueTypes::Yes,
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/fulfill.rs:634]: This should never matter -- const types can't name TAITs (afaict), at least not without more features enabled. This can stay `Yes` but needs a FIXME.

File: compiler/rustc_trait_selection/src/traits/fulfill.rs [hunk 21]
```diff
@@ -589,7 +589,7 @@ impl<'a, 'tcx> ObligationProcessor for FulfillProcessor<'a, 'tcx> {
                     match (evaluate(c1), evaluate(c2)) {
                         (Ok(c1), Ok(c2)) => {
                             match self.selcx.infcx.at(&obligation.cause, obligation.param_env).eq(
-                                DefineOpaqueTypes::No,
+                                DefineOpaqueTypes::Yes,
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/fulfill.rs:592]: Same as the comment below, consts can't have tait types I think.

File: compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs [hunk 22]
```diff
@@ -3592,7 +3592,7 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
                     for (expected, actual) in zipped {
                         self.probe(|_| {
                             match self.at(&ObligationCause::misc(expr.span, body_id), param_env).eq(
-                                DefineOpaqueTypes::No,
+                                DefineOpaqueTypes::Yes,
```
Comment (compiler-errors) [inline compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:3595]: This is diagnostics code. Please leave a comment saying that and that `Yes` doesn't matter.

File: compiler/rustc_hir_typeck/src/method/probe.rs [hunk 23]
```diff
@@ -1485,6 +1485,10 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
         self.probe(|_| {
             // First check that the self type can be related.
             let sub_obligations = match self.at(&ObligationCause::dummy(), self.param_env).sup(
+                // https://github.com/rust-lang/rust/issues/116652
+                // Failed tests:
+                // tests/ui/impl-trait/issues/issue-70877.rs
+                // tests/ui/type-alias-impl-trait/type-alias-impl-trait.rs
```
Comment (lqd) [inline compiler/rustc_hir_typeck/src/method/probe.rs:1491]: We may want to mention that these 2 are ICEs and describe them here?

**Comment by chansuke** [comment]
@compiler-errors Thank you for your comment. I apologize for the misunderstanding regarding the issue. I have reverted the previous commit and made a new commit with the single file change, replacing single `DefineOpaqueTypes::No` that passes the test.

File: compiler/rustc_hir_typeck/src/method/probe.rs (hunk 23 above)
Comment (chansuke) [reply compiler/rustc_hir_typeck/src/method/probe.rs]: @lqd Thank you for your review. Yes, I've reverted and made a new commit

**Comment by chansuke** [comment]
@compiler-errors 
I squashed the commits and added FIXME.

**Comment by Dylan-DPC** [comment]
@chansuke if you can resolve the latest conflicts, we can push this ahead for a review. Thanks

**Comment by compiler-errors** [comment]
I think @oli-obk is picking up this work as part of ATPIT stabilization.
